src/vmm/src/builder.rs

```diff
@@ -510,21 +520,63 @@ pub fn build_microvm_from_snapshot(
     )?;
 
     #[cfg(target_arch = "x86_64")]
-    // Check if we need to scale the TSC.
-    // We start by checking if the CPU model in the snapshot is
-    // the same as this host's. If they are the same, we don't
-    // need to do anything else.
-    if !is_same_model(&microvm_state.vcpu_states[0].cpuid) {
-        // Extract the TSC freq from the state.
-        // No TSC freq in snapshot means we have to fail-fast.
-        let state_tsc = microvm_state.vcpu_states[0]
-            .tsc_khz
-            .ok_or(BuildMicrovmFromSnapshotError::TscFrequencyNotPresent)?;
-
-        // Scale the TSC frequency for all VCPUs, if needed.
-        if vcpus[0].kvm_vcpu.is_tsc_scaling_required(state_tsc)? {
-            for vcpu in &vcpus {
-                vcpu.kvm_vcpu.set_tsc_khz(state_tsc)?;
+    {
+        // Convert CPUID to better format
+        let snapshot_cpuid = {
+            let raw_snapshot_cpuid =
+                cpuid::RawCpuid::from(microvm_state.vcpu_states[0].cpuid.clone());
+
+            // TODO Prevent `microvm_state.vcpu_states[0].cpuid` containing duplicate and invalid
+            // leaves. Do not save snapshots with invalid leaves.
```

Comment:
Something is confusing here; you are calling this at resume time (that is when you restore from a snapshot) but ur comment refers to saving snapshots.

Reply:
This has been moved to https://github.com/firecracker-microvm/firecracker/pull/3360.